examples/distributed/pyg/distributed_cpu.py

```diff
@@ -0,0 +1,497 @@
+            f'partition{node_rank}.pt',
+        ))
+
+    if is_hetero:
+        train_idx = ('paper', train_idx)
+        test_idx = ('paper', test_idx)
+
+    # load partition information
+    (
+        meta,
+        num_partitions,
+        partition_idx,
+        node_pb,
+        edge_pb,
+    ) = load_partition_info(osp.join(root_dir, f'{dataset}-partitions'),
+                            node_rank)
+    print(f'meta={meta}, partition_idx={partition_idx}')
+    # load partition into graph
+    graph = LocalGraphStore.from_partition(
+        osp.join(root_dir, f'{dataset}-partitions'), node_rank)
+    # load partition into feature
+    feature = LocalFeatureStore.from_partition(
+        osp.join(root_dir, f'{dataset}-partitions'), node_rank)
+
+    # setup the partition information in LocalGraphStore and LocalFeatureStore
+    graph.num_partitions = feature.num_partitions = num_partitions
+    graph.partition_idx = feature.partition_idx = partition_idx
+    graph.node_pb = feature.node_feat_pb = node_pb
+    graph.edge_pb = feature.edge_feat_pb = edge_pb
+    graph.meta = feature.meta = meta
+    feature.labels = torch.load(node_label_file)
+    partition_data = (feature, graph)
+
+    # Initialize distributed context
+    current_ctx = DistContext(
+        world_size=num_nodes,
+        rank=node_rank,
+        global_world_size=num_nodes,
+        global_rank=node_rank,
+        group_name='distributed-sage-supervised-Node',
+    )
+    current_device = torch.device('cpu')
+
+    print('--- Initialize DDP training group ...')
+    # Initialize DDP training process group.
+    torch.distributed.init_process_group(
+        backend='gloo',
+        rank=current_ctx.rank,
+        world_size=current_ctx.world_size,
+        init_method='tcp://{}:{}'.format(master_addr, ddp_port),
+    )
+    num_neighbors = [int(i) for i in num_neighbors.split(',')]
+    persistent_workers = (True if num_workers > 0 else False
+                          )  # Keep workers RPC alive outside the iterator loop
+    print('--- Initialize distributed loaders ...')
+    # Create distributed neighbor loader for training
+    train_loader = pyg_dist.DistNeighborLoader(
+        data=partition_data,
+        input_nodes=train_idx,
+        current_ctx=current_ctx,
+        device=current_device,
+        num_neighbors=num_neighbors,
+        shuffle=True,
+        drop_last=True,
+        persistent_workers=persistent_workers,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        master_addr=master_addr,
+        master_port=train_loader_port,
+        concurrency=concurrency,
+        async_sampling=async_sampling,
+    )
+    # Create distributed neighbor loader for testing.
+    test_loader = pyg_dist.DistNeighborLoader(
+        data=partition_data,
+        input_nodes=test_idx,
+        current_ctx=current_ctx,
+        device=current_device,
+        num_neighbors=num_neighbors,
+        shuffle=False,
+        drop_last=True,
```

Comment:
Is this intended?

Reply:
we need to specify `drop_last=True`, shuffle is there just for consistency of input with the train loader above.

Reply:
Is `drop_last` a restriction of distributed training? In the end, this will make the final outputs (and their metrics) non-complete.

Reply:
We observed that when the last batch in partitioned datasets is incomplete it causes adverse gradient update in the training. But you might be right that `drop_last` shouldn't be on for the test. Sorry I missed that.